plugins/out_s3/s3.c

```diff
@@ -2117,17 +2132,35 @@ static void cb_s3_flush(struct flb_event_chunk *event_chunk,
 
     /* Process chunk */
     if (ctx->log_key) {
-        chunk = flb_pack_msgpack_extract_log_key(ctx,
-                                                 event_chunk->data,
-                                                 event_chunk->size);
+        chunk_tmp = flb_pack_msgpack_extract_log_key(ctx,
+                                                     event_chunk->data,
+                                                     event_chunk->size);
     }
     else {
-        chunk = flb_pack_msgpack_to_json_format(event_chunk->data,
-                                                event_chunk->size,
+        chunk_tmp = flb_sds_create_len(event_chunk->data, event_chunk->size);
+
+    }
+
+    if (ctx->format == S3_FORMAT_JSON) {
+        chunk = flb_pack_msgpack_to_json_format(chunk_tmp,
+                                                flb_sds_len(chunk_tmp),
                                                 FLB_PACK_JSON_FORMAT_LINES,
                                                 ctx->json_date_format,
                                                 ctx->date_key);
+        flb_sds_destroy(chunk_tmp);
+    }
+    else if (ctx->format == S3_FORMAT_CSV) {
+        /*
+         * CSV is a bit tricky, if the user asked to include the column names
+         * we need to know if the content will be placed in a new object or not,
+         * for hence the formatting of the CSV will happen inside the function
+         * s3_store_buffer_put().
```

Comment:
I think we can determine in this cb_flush whether or not we are creating a new buffer file or not. 


`s3_store_file_get` returns NULL if no existing buffer file exists that we can use. See the `buffer_chunk` call which is the case where we create a new buffer. 

Also, this needs to work with the fact that S3 can upload data without first buffering it. If the new data + any old data is larger than our upload size, we just upload directly. And so we need to handle CSV in that case too. 

I think this should not be implemented in S3 store consequently.

Reply:
@PettitWesley I've refactor the formatter and simplified the logic, now everything is on s3.c

would you please elaborate on the direct upload stuff ?, I got confused with this approach, does it means a previous content with the new incoming one are sent together ? if so, the logic for column headers should work since a file existed previously so no headers were created.